python/cuml/test/test_kernel_density.py

```diff
@@ -0,0 +1,140 @@
+    return norm*np.average(k, -1, sample_weights)
+
+
+@st.composite
+def array_strategy(draw):
+    n = draw(st.integers(1, 100))
+    m = draw(st.integers(1, 100))
+    dtype = draw(st.sampled_from([np.float64, np.float32]))
+    rng = np.random.RandomState(34)
+    X = rng.randn(n, m).astype(dtype)
+    n_test = draw(st.integers(1, 100))
+    X_test = rng.randn(n_test, m).astype(dtype)
+
+    if draw(st.booleans()):
+        sample_weights = None
+    else:
+        sample_weights = draw(arrays(dtype=np.float64, shape=n,
+                                     elements=st.floats(0.1, 2.0),))
+
+    return X, X_test, sample_weights
+
+
+metrics_strategy = st.sampled_from(
+    ['euclidean', 'manhattan',
+     'chebyshev', 'minkowski', 'hamming', 'canberra'])
+
+
+@settings(deadline=None, max_examples=100)
+@given(array_strategy(), st.sampled_from(VALID_KERNELS),
+       metrics_strategy, st.floats(0.2, 10))
+def test_kernel_density(arrays, kernel, metric, bandwidth):
+    X, X_test, sample_weights = arrays
+    if kernel == 'cosine':
+        # cosine is numerically unstable at high dimensions
+        # for both cuml and sklearn
+        assume(X.shape[1] <= 20)
+    kde = KernelDensity(kernel=kernel, metric=metric,
+                        bandwidth=bandwidth).fit(X,
+                                                 sample_weight=sample_weights)
+    cuml_prob = kde.score_samples(X)
+    cuml_prob_test = kde.score_samples(X_test)
+
+    if X.dtype == np.float64:
+        ref_prob = compute_kernel_naive(
+            X, X, kernel, metric, bandwidth, sample_weights)
+        ref_prob_test = compute_kernel_naive(
+            X_test, X, kernel, metric, bandwidth, sample_weights)
+        tol = 1e-3
+        assert np.allclose(np.exp(cuml_prob), ref_prob,
+                           rtol=tol, atol=tol, equal_nan=True)
+        assert np.allclose(np.exp(cuml_prob_test),
+                           ref_prob_test, rtol=tol, atol=tol, equal_nan=True)
+
+    if kernel in ["gaussian", "tophat"] and metric == "euclidian":
+        sample = kde.sample(100, random_state=32).get()
+        nearest = skl_pairwise_distances(sample, X, metric=metric)
+        nearest = nearest.min(axis=1)
+        if kernel == "gaussian":
+            assert np.all(nearest < 5 * bandwidth)
+        elif kernel == "tophat":
+            assert np.all(nearest <= bandwidth)
+    else:
+        with pytest.raises(NotImplementedError):
+            kde.sample(100)
+
+
+def test_logaddexp():
+    X = np.array([[0.0, 0.0], [0.0, 0.0]])
+    out = np.zeros(X.shape[0])
+    logsumexp_kernel.forall(out.size)(X, out)
+    assert np.allclose(out, np.logaddexp.reduce(X, axis=1))
+
+    X = np.array([[3.0, 1.0], [0.2, 0.7]])
+    logsumexp_kernel.forall(out.size)(X, out)
+    assert np.allclose(out, np.logaddexp.reduce(X, axis=1))
+
+
+@pytest.mark.xfail(
+    reason="cuml's pairwise_distances does"
+    "not process metric_params as expected")
```

Comment:
The only use-case for metric_params is `{'p': X}` for the `minkowski` distance? If so it can be a special case in the KDE code, or you could open an issue to have this behavior implemented (so that this xfail is tracked)

Reply:
I addressed this by attempting to turn the dict input into a scalar and pass to pairwise_distances. Ideally pairwise_distances would accept a dict like sklearn and do some error checking but I don't want to deal with that here.